README.md
# Soul Engine fnf
this engine is made for android
Made to make fnf mods port.








:)


CHANGELOG.md
#Version
#Alpha 0.10.25
-new notes
-new difficulty "noob"
#Alpha 0.5
-New difficulty "Crazy"
-New Menu Position
-New story mode menu with "purple" color
#Alpha 0.3
None


assets/preload/stages/eddhouse.lua
function onCreate()
	precacheImage('DoorOpen')
	precacheImage('Matt')
	precacheImage('Tom')
	precacheImage('Eduardo')
	precacheImage('EduardoCUTSCENE')
	precacheImage('Mark')
	precacheImage('John')
	precacheImage('character/EddSide')
	precacheImage('character/DEdd')
	makeAnimationList();
	-- background shit
	makeLuaSprite('sky', 'SkyBox', -1790, -800);
	setScrollFactor('sky', 0.1, 0.5);
	
	makeLuaSprite('cloud', 'Clouds', -2590, -500);
	setScrollFactor('cloud', 0.1, 0.3);
	setProperty('cloud.velocity.x', math.random(5, 15));
	
	makeLuaSprite('houses', 'HousesAndFloor', -1790, -600);
	setScrollFactor('houses', 1, 1);
	
	if not lowQuality then
		makeAnimatedLuaSprite('doorOpen', 'DoorOpen', 600, 300); -- from that door on Hard comes up Matt with Tom
		addAnimationByPrefix('doorOpen', 'open', 'Door Opening', 12, false)
		setScrollFactor('doorOpen', 1, 1);
		setProperty('doorOpen.alpha', 0);
		scaleObject('doorOpen', 1.3, 1.3);
	
		makeLuaSprite('city', 'SecondParalax', -1290, -500);
		setScrollFactor('city', 0.65, 0.65);
		scaleObject('city', 0.8, 0.8);
		
		makeAnimatedLuaSprite('matt', 'Matt', 560, 245); -- matt !!
		addAnimationByPrefix('matt', 'walk', 'MattWalking', 24, true)
		addAnimationByPrefix('matt', 'idle', 'MattSnappingFinger', 24, false)
		addAnimationByPrefix('matt', 'reaction', 'MattReeaction', 24, false)
		addAnimationByPrefix('matt', 'idol', 'MattPISSED', 24, false)
		setScrollFactor('matt', 1, 1);
		setProperty('matt.alpha', 0);
		scaleObject('matt', 1.5, 1.5);
		
		makeAnimatedLuaSprite('tom', 'Tom', 610, 245); -- tom !!
		addAnimationByPrefix('tom', 'walk', 'TomWalkingBy', 24, true)
		addAnimationByPrefix('tom', 'trans', 'TomTransition', 24, false)
		addAnimationByPrefix('tom', 'idle', 'TomIdle', 24, true)
		addAnimationByPrefix('tom', 'lookin', 'TomLooking', 24, false)
		addAnimationByPrefix('tom', 'idol', 'TomIdol', 24, false)
		addAnimationByPrefix('tom', 'reaction', 'TomReaction', 24, false)
		setScrollFactor('tom', 1, 1);
		setProperty('tom.alpha', 0);
		scaleObject('tom', 1.5, 1.5);
		
		makeAnimatedLuaSprite('john', 'John', -915, 205); -- john
		addAnimationByPrefix('john', 'idle', 'JohnIdle', 24, false)
		setScrollFactor('john', 1, 1);
		setProperty('john.alpha', 0);
		scaleObject('john', 0.9, 0.9);
		
		makeAnimatedLuaSprite('mark', 'Mark', -760, 245); -- mark
		addAnimationByPrefix('mark', 'idle', 'MarkIdle', 24, false)
		setScrollFactor('mark', 1, 1);
		setProperty('mark.alpha', 0);
		scaleObject('mark', 0.8, 0.8);
		
		makeLuaSprite('car', 'Car', -1790, -600);
		setScrollFactor('car', 1.15, 1.15);
		
		makeLuaSprite('plane', 'Plane', -890, 0);
		setScrollFactor('plane', 0.2, 0.6);
	end
	
	makeLuaSprite('fence', 'Fence', -1790, -600);
	setScrollFactor('fence', 1, 1);
		
	makeAnimatedLuaSprite('eduarC', 'EduardoCUTSCENE', -1080, 125); -- eduardo punches john, cry about it =l
	addAnimationByPrefix('eduarC', 'punch', 'EduardoPunch', 24, false)
	setScrollFactor('eduarC', 1, 1);
	setProperty('eduarC.alpha', 0);
		
	makeAnimatedLuaSprite('eduar', 'Eduardo', -1080, 125); -- eduardo
	addAnimationByPrefix('eduar', 'idle', 'EduardoIdle', 24, false)
	addAnimationByPrefix('eduar', 'singLEFT', 'EduardoLeft', 24, false)
	addAnimationByPrefix('eduar', 'singDOWN', 'EduardoDown', 24, false)
	addAnimationByPrefix('eduar', 'singUP', 'EduardoUp', 24, false)
	addAnimationByPrefix('eduar', 'singRIGHT', 'EduardoRight', 24, false)
	addAnimationByPrefix('eduar', 'well', 'EduardoWell', 24, false)
	setScrollFactor('eduar', 1, 1);
	setProperty('eduar.alpha', 0);

	addLuaSprite('sky', false); --bg
	addLuaSprite('cloud', false);
	addLuaSprite('plane', false);
	addLuaSprite('city', false);
	addLuaSprite('houses', false);
	addLuaSprite('doorOpen', false);
	
	addLuaSprite('mark', false); --dumbasses
	addLuaSprite('john', false);
	addLuaSprite('eduar', false);
	addLuaSprite('eduarC', false);
	addLuaSprite('fence', false); --fence lol
	
	addLuaSprite('matt', false); -- matt
	addLuaSprite('tom', false); --tom
	
	addLuaSprite('car', true);
end

function onStepHit()
	if curStep == 144 then -- tord plane comin
		doTweenX('PlaneTweenX', 'plane', 1600, 25);
	end
	if curStep == 272 then -- matt coming up
		objectPlayAnimation('doorOpen', 'open', true)
		setProperty('doorOpen.alpha', 1);
		runTimer('doorGoesInvis', 1, 1)
		setProperty('matt.alpha', 1);
		doTweenX('MattTweenX', 'matt', 10, 1.9);
		function onTimerCompleted(tag, loops, loopsLeft)
			if tag == 'doorGoesInvis' then
				setProperty('doorOpen.alpha', 0);
			end
		end
	end
	if curStep == 416 then -- tom coming up
		setProperty('doorOpen.alpha', 1);
		objectPlayAnimation('doorOpen', 'open', true)
		runTimer('doorGoesInvis', 1, 1)
		setProperty('tom.alpha', 1);
		setProperty('tom.flipX', true);
		doTweenX('tomTweenX', 'tom', 1390, 4.8);
		function onTimerCompleted(tag, loops, loopsLeft)
			if tag == 'doorGoesInvis' then
				setProperty('doorOpen.alpha', 0);
			end
		end
	end
	if curStep == 928 then -- well
		setProperty('eddI.visible', true)
		objectPlayAnimation('eduar', 'well', true)
		doTweenAlpha('SkyTweenAlpha', 'sky', 0.9, 0.2);
		doTweenZoom('ZoominToNeighbours', 'camGame', 0.75, 0.2);
	end
	if curStep == 932 then -- well
		objectPlayAnimation('eduar', 'well', true)
		doTweenAlpha('SkyTweenAlpha2', 'sky', 0.8, 0.2);
		doTweenZoom('ZoominToNeighbours2', 'camGame', 0.8, 0.2);
	end
	if curStep == 936 then -- well
		objectPlayAnimation('eduar', 'well', true)
		doTweenAlpha('SkyTweenAlpha3', 'sky', 0.7, 0.2);
		doTweenZoom('ZoominToNeighbours3', 'camGame', 0.9, 0.2);
		runTimer('SkyToNormal', 1, 1)
		function onTimerCompleted(tag, loops, loopsLeft)
			if tag == 'SkyToNormal' then
				doTweenAlpha('SkyTweenAlpha4', 'sky', 1, 2);
			end
		end
	end
	if curStep == 1599 then -- but i didnt even said anything
		removeLuaSprite('mark')
		removeLuaSprite('eduar')
		removeLuaSprite('john')
		setProperty('eduarC.alpha', 1);
		objectPlayAnimation('eduarC', 'punch', true)
	end
end -- hello im serdzhant

function onBeatHit()
	if curBeat % 2 == 0 and curStep > 289 and curStep < 912 then
		objectPlayAnimation('matt', 'idle', true);
	end
	if curBeat % 2 == 0 and curStep > 478 then
		objectPlayAnimation('tom', 'idle', true);
	end
	if curBeat == 228 then
		doTweenAlpha('hudFunne', 'camHUD', 0, 0.6, 'linear')
		objectPlayAnimation('tom', 'reaction', true);
		objectPlayAnimation('matt', 'reaction', true);
		setProperty('matt.x', -80); -- offset
		setProperty('eduar.alpha', 1);
		setProperty('john.alpha', 1);
		setProperty('mark.alpha', 1);
	end
	if curBeat % 2 == 0 and curStep > 940 then
		objectPlayAnimation('matt', 'idol', true);
		setProperty('matt.x', -120); -- offset
		objectPlayAnimation('tom', 'idol', true);
		objectPlayAnimation('eduar', 'idle', true);
		objectPlayAnimation('john', 'idle', true);
		objectPlayAnimation('mark', 'idle', true);
	end
	if curBeat == 236 then
		doTweenAlpha('hudFunne', 'camHUD', 1, 0.2, 'linear')
	end
	if curBeat == 398 then -- cutscene ending
		doTweenAlpha('hudFunne', 'camHUD', 0, 0.4, 'linear') -- ALEX DIRECTOR WROTE THIS WOOO-
	end
end
--eduardo shit
--script by Shadow Mario, Kevin Kuntz and other dudos
animationsList = {}
holdTimers = {eduar = -1.0};
noteDatas = {eduar = 0};
function makeAnimationList()
	animationsList[0] = 'singLEFT';
	animationsList[1] = 'singDOWN';
	animationsList[2] = 'singUP';
	animationsList[3] = 'singRIGHT';
end

function goodNoteHit(id, direction, noteType, isSustainNote)
	if boyfriendName == 'bf' then
		setPropertyFromClass('GameOverSubstate', 'characterName', 'bf');
	end
	if noteType == 'Opponent Sing' then
		setPropertyFromClass('GameOverSubstate', 'characterName', 'edd-dead');
	end
end
function opponentNoteHit(id, direction, noteType, isSustainNote)
	if noteType == 'Special Sing' then
		if not isSustainNote then
			noteDatas.eduar = direction;
		end	
		characterToPlay = 'eduar'
		animToPlay = noteDatas.eduar;
			
		playAnimation(characterToPlay, animToPlay, true);
	end
end

function onUpdate(elapsed)
	if getProperty('dad.curCharacter') == 'eddside' then
		if getProperty('dad.animation.curAnim.name') == 'what' then
			setProperty('dad.holdTimer', 0);
		end
	end
end

function playAnimation(character, animId, forced)
	animName = animationsList[animId];
	--debugPrint(animName);
	if character == 'eduar' then
		objectPlayAnimation('eduar', animName, forced); -- this part is easily broke if you use objectPlayAnim (I have no idea why its like this)
	end
end
--eduardo shit ends

function onTweenCompleted(tag)
	if tag == 'PlaneTweenX' then
		removeLuaSprite('plane', true); -- optimization
	end
	if tag == 'MattTweenX' then
		objectPlayAnimation('matt', 'idle', true)
		setProperty('matt.x', -120); -- offset
	end
	if tag == 'tomTweenX' then
		objectPlayAnimation('tom', 'trans', true);
		setProperty('tom.flipX', false);
	end
end

function onCreatePost()
		makeLuaSprite('eddI', 'edd', getProperty('iconP1.x'), getProperty('iconP1.y'))
		setObjectCamera('eddI', 'hud')
		addLuaSprite('eddI', true)
		setObjectOrder('eddI', getObjectOrder('iconP1') + 1)
		setProperty('eddI.flipX', true)
		setProperty('eddI.visible', false)
end

function onUpdatePost(elapsed)
		setProperty('eddI.x', getProperty('iconP1.x') + 50)
		setProperty('eddI.angle', getProperty('iconP1.angle'))
		setProperty('eddI.y', getProperty('iconP1.y') - 50)
		setProperty('eddI.scale.x', getProperty('iconP1.scale.x') - 0.3)
		setProperty('eddI.scale.y', getProperty('iconP1.scale.y') - 0.3)
end

assets/preload/stages/eddhouse-tord.lua
function onCreate()
	precacheImage('DoorOpen')
	precacheImage('Matt')
	precacheImage('Fallin')
	precacheImage('character/Tord')
	precacheImage('character/Tord2')
	precacheImage('character/SlideEdd')
	precacheImage('character/SlideBF')
	precacheImage('character/EddTordCutscene')
	precacheImage('character/BFTordCutscene')
	precacheImage('character/DEdd')
	precacheImage('notes/Explosion')
	-- background shit
	makeLuaSprite('sky', 'SkyBox', -1790, -800);
	setScrollFactor('sky', 0.1, 0.5);
	
	makeLuaSprite('cloud', 'Clouds', -2590, -500);
	setScrollFactor('cloud', 0.1, 0.3);
	setProperty('cloud.velocity.x', math.random(5, 15));
	
	makeLuaSprite('houses', 'HousesAndFloor', -1790, -600);
	setScrollFactor('houses', 1, 1);
	
	if not lowQuality then
		makeAnimatedLuaSprite('doorOpen', 'DoorOpen', 600, 300); -- from that door on Hard comes up Matt with Tom
		addAnimationByPrefix('doorOpen', 'open', 'Door Opening', 12, false)
		setScrollFactor('doorOpen', 1, 1);
		setProperty('doorOpen.alpha', 0);
		scaleObject('doorOpen', 1.3, 1.3);
	
		makeLuaSprite('city', 'SecondParalax', -1390, -500);
		setScrollFactor('city', 0.65, 0.65);
		scaleObject('city', 0.9, 0.8);
		
		makeAnimatedLuaSprite('matt', 'Matt', 560, 245); -- matt !!
		addAnimationByPrefix('matt', 'walk', 'MattWalking', 24, true)
		addAnimationByPrefix('matt', 'idle', 'MattSnappingFinger', 24, false)
		addAnimationByPrefix('matt', 'reaction', 'MattReactionTord', 24, false)
		addAnimationByPrefix('matt', 'lookin', 'MattHarpoonBit', 24, false)
		addAnimationByPrefix('matt', 'lookidle', 'MattHarpoonIdle', 24, true)
		setScrollFactor('matt', 1, 1);
		setProperty('matt.alpha', 0);
		scaleObject('matt', 1.5, 1.5);
		
		makeAnimatedLuaSprite('tom', 'TomEND', 1210, 299); -- tom !!
		addAnimationByPrefix('tom', 'walk', 'Tom Running In', 24, false)
		addAnimationByPrefix('tom', 'harp', 'TomHarpoonLine', 24, false)
		addAnimationByPrefix('tom', 'harpidle', 'TomHarpoonIdle', 24, true)
		setScrollFactor('tom', 1, 1);
		setProperty('tom.alpha', 0);
		scaleObject('tom', 1.5, 1.5);
		
		makeLuaSprite('car', 'Car', -1790, -600);
		setScrollFactor('car', 1.15, 1.15);
		
		makeLuaSprite('plane', 'Plane', -890, 0);
		setScrollFactor('plane', 0.2, 0.6);
		
		makeAnimatedLuaSprite('tord', 'TordHelicopter', 1270, -45); -- this red mf
		addAnimationByIndices('tord', 'idle', 'TordHelicopter', '1, 2', 12)
		addAnimationByPrefix('tord', 'mama', 'TordHelicopter', 24, false)
		setScrollFactor('tord', 0.2, 0.6);
		setProperty('tord.alpha', 0);
		scaleObject('tord', 2, 2);
		
		makeAnimatedLuaSprite('tord2', 'TordFallin', 600, -845); -- this red mf fallin
		addAnimationByPrefix('tord2', 'idle', 'TordFlailing', 24, true)
		setScrollFactor('tord2', 0.9, 0.9);
		setProperty('tord2.alpha', 0);
		scaleObject('tord2', 1.6, 1.6);
		
		makeAnimatedLuaSprite('expl', 'notes/Explosion', 760, -110); -- explosion after rocket note
		addAnimationByPrefix('expl', 'idle', 'Explosion', 20, false)
		setScrollFactor('expl', 0, 0);
		setProperty('expl.alpha', 0);
		scaleObject('expl', 1.5, 1.5);
		setObjectCamera('expl', 'other')
		
		makeAnimatedLuaSprite('targ', 'notes/Target', 957, 50); -- spawns by Rocket Alert Event
		addAnimationByPrefix('targ', 'idle', 'Target', 20, false)
		setScrollFactor('targ', 0, 0);
		setProperty('targ.alpha', 0);
		scaleObject('targ', 0.6, 0.6);
		setObjectCamera('targ', 'other')
		
		if downscroll and not middlescroll then
			setProperty('expl.y', 350);
			setProperty('targ.y', 570);
		end
		if downscroll and middlescroll then
			setProperty('expl.y', 350);
			setProperty('expl.x', 430);
			setProperty('targ.y', 570);
			setProperty('targ.x', 640);
		end
		if not downscroll and middlescroll then
			setProperty('expl.x', 430);
			setProperty('targ.x', 640);
		end
	end
		
	makeAnimatedLuaSprite('tordbot', 'TordBot', 340, -145); -- this red mf
	addAnimationByPrefix('tordbot', 'idle', 'TordBot0', 24, true)
	addAnimationByPrefix('tordbot', 'explode', 'TordBotBlowingUp', 24, true)
	addAnimationByIndices('tordbot', 'blowidle', 'TordBotBlowingUp', '1, 2', 24)
	setScrollFactor('tordbot', 0.9, 0.9);
	scaleObject('tordbot', 1.3, 1.3);
	
	makeLuaSprite('fence', 'Fence', -1790, -600);
	setScrollFactor('fence', 1, 1);
	
	makeLuaSprite('tordBG', 'TordBG', -100, -200);
	setScrollFactor('tordBG', 0, 0);
	scaleObject('tordBG', 0.5, 0.5);
	setProperty('tordBG.alpha', 0);
	
	makeLuaSprite('cock', 'CockPitUpClose', -20, -40);
	setScrollFactor('cock', 0, 0);
	scaleObject('cock', 0.6, 0.6);
	setProperty('cock.alpha', 0);

	addLuaSprite('sky', false); --bg
	addLuaSprite('cloud', false);
	addLuaSprite('plane', false);
	addLuaSprite('tord', false);
	addLuaSprite('city', false);
	addLuaSprite('tordbot', false);
	addLuaSprite('tord2', false);
	addLuaSprite('houses', false);
	addLuaSprite('doorOpen', false);
	
	addLuaSprite('fence', false); --fence lol
	
	addLuaSprite('matt', false); -- matt
	addLuaSprite('tom', true); --tom
	
	addLuaSprite('car', true);
	addLuaSprite('tordBG', false);
	addLuaSprite('cock', true);
	
	addLuaSprite('expl', true);
	addLuaSprite('targ', true);
end

function onStepHit()
	if curStep == 144 then -- tord plane comin
		doTweenX('PlaneTweenX', 'plane', 1600, 25);
	end
	if curStep == 272 then -- matt coming up
		objectPlayAnimation('doorOpen', 'open', true)
		setProperty('doorOpen.alpha', 1);
		runTimer('doorGoesInvis', 1, 1)
		setProperty('matt.alpha', 1);
		doTweenX('MattTweenX', 'matt', 10, 1.9);
	end
	if curStep == 397 then -- tord comnin
		doTweenY('TordTweenY', 'tord', 110, 2, 'cubeIn');
		objectPlayAnimation('tord', 'idle', true)
		setProperty('tord.alpha', 1);
	end
	if curStep == 928 then -- tordbot comnin
		doTweenAlpha('hudFunne', 'camHUD', 0, 0.2, 'linear')
		objectPlayAnimation('matt', 'reaction', true);
		setProperty('cameraSpeed', getProperty('cameraSpeed') - 0.8)
		doTweenY('TordbotTweenY', 'tordbot', -615, 5.1);
	end
	if curStep == 936 then -- tordbot comnin
		setProperty('tom.alpha', 1);
		objectPlayAnimation('tom', 'walk', true);
	end
	if curStep == 1020 then -- tordbot comnin
		doTweenZoom('FunnyZoomin', 'camGame', 1.3, 0.7, 'cubeOut');
		setProperty('defaultCamZoom', 1.3);
	end
	if curStep == 1024 then -- inside tordbot intro
		setProperty('eddI.visible', true)
		doTweenAlpha('hudFunne', 'camHUD', 1, 1, 'linear')
		setProperty('cameraSpeed', 3);
		setProperty('cock.alpha', 1);
		doTweenAlpha('cockTweenalfa', 'cock', 0, 0.7);
		setProperty('tordBG.alpha', 1);
		setProperty('gf.y', -900); -- tord as gf
		setProperty('gf.x', 100); -- tord as gf
	end
	if curStep == 1132 then -- slide edd
		setProperty('dad.y', -100);
		setProperty('dad.x', 50);
	end
	if curStep == 1260 then -- slide bf
		setProperty('boyfriend.y', -300);
		setProperty('boyfriend.x', 1150);
	end
	if curStep == 1312 then -- offsets reason of cuz 'change character' event changing it
		setProperty('gf.y', -925); -- tord as gf
		setProperty('gf.x', 85); -- tord as gf
	end
	if curStep == 1344 then -- offsets reason of cuz 'change character' event changing it
		setProperty('gf.y', -900); -- tord as gf
		setProperty('gf.x', 100); -- tord as gf
	end
	if curStep == 1984 then -- offsets reason of cuz 'change character' event changing it
		doTweenAlpha('hudFunne', 'camHUD', 0, 0.2, 'linear')
		setProperty('gf.y', -925); -- tord as gf
		setProperty('gf.x', 85); -- tord as gf
	end
	if curStep == 2016 then -- camera getin out of tordbot
		objectPlayAnimation('tordbot', 'blowidle', true)
		setProperty('tordbot.x', -220); -- animation offset
		setProperty('tordbot.y', -880); -- animation offset
		doTweenX('tordbotShakeX', 'tordbot', -230, 0.1);
		setProperty('gf.y', 130); -- gf comin home
		setProperty('gf.x', 460); -- gf comin home
		removeLuaSprite('tordBG'); --optimization!!
		doTweenZoom('FunnyZoomin', 'camGame', 0.7, 0.3);
		setProperty('defaultCamZoom', 0.7);
		setProperty('cameraSpeed', getProperty('cameraSpeed') - 0.4);
		setProperty('cock.alpha', 1);
		doTweenAlpha('cockTweenalfa', 'cock', 0, 0.3);
	end
	if curStep == 2032 then -- KABOOOM
		runTimer('tordbotExploded', 0.7, 1)
		objectPlayAnimation('tordbot', 'explode', true)
		setProperty('tom.x', 700); --offset
		setProperty('matt.x', 100); --offset
		setProperty('cameraSpeed', 0.2);
	end
	if curStep == 2047 then -- camera slow movin shit
		objectPlayAnimation('tom', 'harpidle', true)
		objectPlayAnimation('matt', 'lookidle', true)
		setProperty('tord2.alpha', 1);
		doTweenY('tordFallinAgainTweenY', 'tord2', 400, 1.5, 'linear');
	end
	if curStep == 2078 then -- tom sayin funny wors
		objectPlayAnimation('matt', 'lookin', true)
		setProperty('matt.x', 35); --offset
	end
	if curStep == 2087 then -- tom sayin funny wors
		objectPlayAnimation('tom', 'harp', true)
	end
end -- hello im serdzhant

function onBeatHit()
	if curBeat % 2 == 0 and curStep > 292 and curStep < 929 then
		objectPlayAnimation('matt', 'idle', true);
	end
end

function goodNoteHit(id, direction, noteType, isSustainNote) --slide bf and slide edd inside tordbot
	if noteType == 'Special Sing' and boyfriendName == 'slide-bf' then
		setPropertyFromClass('GameOverSubstate', 'characterName', 'bf');
		doTweenY('BfTweenY', 'boyfriend', -551, 0.1, 'cubeIn');
		doTweenX('BfTweenX', 'boyfriend', 980, 0.1, 'cubeIn');
	end
	if noteType == 'Rocket Note' then
		setProperty('expl.alpha', 1);
		objectPlayAnimation('expl', 'idle', true)
		runTimer('boom', 0.3, 1)
	end
	if noteType == 'Opponent Sing' and dadName == 'slide-edd' then
		setPropertyFromClass('GameOverSubstate', 'characterName', 'edd-dead');
		doTweenY('EddTweenY', 'dad', -470, 0.1, 'cubeIn');
		doTweenX('EddTweenX', 'dad', 230, 0.1, 'cubeIn');
	end
end

function opponentNoteHit(id, direction, noteType, isSustainNote) --health drain by tord
	if gfName == 'tord' then
		triggerEvent('Screen Shake', '0.1, 0.002', '0.1, 0.001');
		setProperty("health", getProperty("health") - 0.03)
		if getProperty("health") < 0.2 then
			setProperty("health", 0.2)
		end
	end
	if noteType == 'Rocket Note' and dadName == 'tord' then
		characterPlayAnim('gf', 'singDOWN-alt', true);
	end
end

function onEvent(f)
	if f == 'Rocket Alert' then
		setProperty('targ.alpha', 1);
		objectPlayAnimation('targ', 'idle', true)
		runTimer('target', 1.2, 1)
	end
end

function onTweenCompleted(tag) 
	if tag == 'PlaneTweenX' then
		removeLuaSprite('plane', true); -- optimization
	end
	if tag == 'MattTweenX' then
		objectPlayAnimation('matt', 'idle', true)
		setProperty('matt.x', -120); -- offset
	end
	if tag == 'TordTweenY' then
		doTweenY('Tord2TweenY', 'tord', 510, 10, 'linear');
		doTweenX('TordGoesLeft', 'tord', 1200, 1.6, 'cubeInOut');
		objectPlayAnimation('tord', 'mama', true)
	end
	if tag == 'TordGoesLeft' then
		doTweenX('TordGoesRight', 'tord', 1400, 1.6, 'cubeInOut');
	end
	if tag == 'TordGoesRight' then
		doTweenX('TordGoesLeft', 'tord', 1200, 1.6, 'cubeInOut');
	end
	if tag == 'Tord2TweenY' then
		removeLuaSprite('tord');  -- optimization
	end
	if tag == 'tordFallinAgainTweenY' then
		removeLuaSprite('tord2');  -- optimization
	end
	if tag == 'tordbotShakeX' and curStep < 2033 then
		doTweenX('tordbotShakinX', 'tordbot', -210, 0.1);
	end
	if tag == 'tordbotShakinX' and curStep < 2033 then
		doTweenX('tordbotShakeX', 'tordbot', -230, 0.1);
	end
	if tag == 'BfTweenX' then
		runTimer('slideBf', 0.2, 1)
	end
	if tag == 'EddTweenX' then
		runTimer('slideEdd', 0.2, 1)
	end
end

function onTimerCompleted(tag, loops, loopsLeft)
	if tag == 'doorGoesInvis' then
		setProperty('doorOpen.alpha', 0);
	end
	if tag == 'tordbotExploded' then
		removeLuaSprite('tordbot'); --optimization
	end
	if tag == 'boom' then
		setProperty('expl.alpha', 0);
	end
	if tag == 'slideEdd' and curStep < 1984 then
		doTweenY('EddTweenY', 'dad', -300, 1, 'cubeIn');
		doTweenX('EddTweenX', 'dad', 0, 1, 'cubeIn');
	end
	if tag == 'slideBf' and curStep < 1984 then
		doTweenY('BfTweenY', 'boyfriend', -300, 1, 'cubeIn');
		doTweenX('BfTweenX', 'boyfriend', 1150, 1, 'cubeIn');
	end
	if tag == 'target' then
		setProperty('targ.alpha', 0);
	end
end

function onCreatePost() -- script by Kevin Kuntz
	makeLuaSprite('eddI', 'edd', getProperty('iconP1.x'), getProperty('iconP1.y'))
	setObjectCamera('eddI', 'hud')
	addLuaSprite('eddI', true)
	setObjectOrder('eddI', getObjectOrder('iconP1') + 1)
	setProperty('eddI.flipX', true)
	setProperty('eddI.visible', false)
end

function onUpdatePost(elapsed)
	setProperty('eddI.x', getProperty('iconP1.x') + 50)
	setProperty('eddI.angle', getProperty('iconP1.angle'))
	setProperty('eddI.y', getProperty('iconP1.y') - 50)
	setProperty('eddI.scale.x', getProperty('iconP1.scale.x') - 0.3)
	setProperty('eddI.scale.y', getProperty('iconP1.scale.y') - 0.3)
end